foursquare-oauth-library/src/main/java/com/foursquare/android/nativeoauth/FoursquareOAuthWebviewActivity.java

```diff
@@ -0,0 +1,283 @@
+                super.onPageStarted(view, url, favicon);
+                Log.d(TAG, "onPageStarted: " + url);
+                setProgressBarIndeterminateVisibility(true);
+
+                if (url.startsWith("https://foursquare.com/oauth2/authenticate?")) {
+                    Uri uri = Uri.parse(url);
+                    Intent result = new Intent();
+                    String denied = uri.getQueryParameter("denied");
+                    if ("1".equals(denied)) {
+                        result.putExtra(FoursquareOAuth.INTENT_RESULT_DENIED, true);
+                        setResult(Activity.RESULT_OK, result);
+                    }
+                }
+            }
+
+            @Override
+            public void onPageFinished(WebView view, String url) {
+                super.onPageFinished(view, url);
+                setProgressBarIndeterminateVisibility(false);
+            }
+
+            @Override
+            public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
+                super.onReceivedError(view, errorCode, description, failingUrl);
+                Intent data = new Intent();
+                data.putExtra(FoursquareOAuth.INTENT_RESULT_ERROR, ERROR_CODE_INTERNAL_ERROR);
+                data.putExtra(FoursquareOAuth.INTENT_RESULT_ERROR_MESSAGE, description);
+                setResult(Activity.RESULT_OK, data);
+                finish();
+            }
+
+            @Override
+            public boolean shouldOverrideUrlLoading(WebView view, String url) {
+                Log.d(TAG, "shouldOverrideUrlLoading: " + url);
+                Uri uri = Uri.parse(url);
+
+                /*
+                 * Check if redirect url is of the form
+                 * foursquareauth://callback?code=CODE&error=ERROR, then extract
+                 * the code or error code from it.
+                 */
+                if (URI_SCHEME.equals(uri.getScheme())
+                        && URI_AUTHORITY.equals(uri.getAuthority())) {
+                    String code = uri.getQueryParameter("code");
+                    String error = uri.getQueryParameter("error");
+                    Intent result = new Intent();
+
+                    if (TextUtils.isEmpty(error)) {
+                        result.putExtra(FoursquareOAuth.INTENT_RESULT_CODE, code);
+
+                    } else {
+                        if (ERROR_CODE_ACCESS_DENIED.equals(error)) {
+                            result.putExtra(FoursquareOAuth.INTENT_RESULT_DENIED, true);
+                        } else {
+                            result.putExtra(FoursquareOAuth.INTENT_RESULT_ERROR, error);
+                        }
+                    }
+
+                    setResult(Activity.RESULT_OK, result);
+                    finish();
+                    return true;
+                }
+                return false;
+            }
+        });
```

Comment:
Can we extract this?

Reply:
technically... yes. With what intent?

Reply:
Simply keeps it easier to read vs having indented code :)

Reply:
all code that isnt import statements are indented :p

Reply:
😂😂😂